FLT/NumberField/InfinitePlace/Extension.lean

```diff
@@ -0,0 +1,334 @@
+image of `f` is real while the image of `g` is complex.
+
+This is the complex embedding analogue of `InfinitePlace.RamifiedExtension`.
+-/
+abbrev IsMixedExtension :=
+  IsExtension f g ∧ ComplexEmbedding.IsReal f ∧ ¬ComplexEmbedding.IsReal g
+
+theorem not_isReal_of_not_isReal {f : K →+* ℂ} {g : L →+* ℂ} (h : IsExtension f g)
+    (hf : ¬ComplexEmbedding.IsReal f) :
+    ¬ComplexEmbedding.IsReal g :=
+  mt (IsReal.comp _) (h ▸ hf)
+
+namespace IsMixedExtension
+
+theorem isExtension (h : IsMixedExtension f g) :
+    IsExtension f g := h.1
+
+theorem isReal (h : IsMixedExtension f g) :
+    ComplexEmbedding.IsReal f := h.2.1
+
+theorem not_isReal (h : IsMixedExtension f g) :
+    ¬ComplexEmbedding.IsReal g := h.2.2
+
+end IsMixedExtension
+
+/--
+If `L/K` and `f : K →+* ℂ`, `g : L →+* ℂ`, then `g` is an _unmixed extension_ of `f` if `g` is an
+extension of `f` but is not a mixed extension. In other words, the image of `f` is real
+if and only if the image of `g` is real.
+
+This is the complex embedding analogue of `InfinitePlace.UnramifiedExtension`.
+-/
+abbrev IsUnmixedExtension := IsExtension f g ∧ ¬IsMixedExtension f g
+
+variable {f g} in
+theorem IsUnmixedExtension.isReal_of_isReal (h : IsUnmixedExtension f g)
+    (hf : ComplexEmbedding.IsReal f) :
+    ComplexEmbedding.IsReal g := by
+  simp only [IsUnmixedExtension, not_and, not_not] at h
+  exact h.2 h.1 hf
+
+/--
+The extensions `g : L →+* ℂ` of `f : K →+* ℂ` are the direct sum of the mixed and the unmixed
+extensions.
+-/
+def isExtensionEquivSum (f : K →+* ℂ) :
+    { g : L →+* ℂ // IsExtension f g } ≃
+      { g : L →+* ℂ // IsMixedExtension f g } ⊕ { g : L →+* ℂ // IsUnmixedExtension f g } :=
+  (Equiv.sumCompl
+    (fun g => ComplexEmbedding.IsReal f ∧ ¬ComplexEmbedding.IsReal g.1)).symm.trans <|
+    Equiv.sumCongr
+      (Equiv.subtypeSubtypeEquivSubtypeInter _ (fun g => _ ∧ ¬ComplexEmbedding.IsReal g))
+      ((Equiv.subtypeSubtypeEquivSubtypeInter _
+        (fun g => ¬(_ ∧ ¬ComplexEmbedding.IsReal g))).trans
+        (Equiv.subtypeEquiv (Equiv.refl _) (fun _ => by aesop)))
+
+end NumberField.ComplexEmbedding
+
+namespace NumberField.InfinitePlace
+
+open NumberField.ComplexEmbedding
+
+variable {K : Type*} {L : Type*} [Field K] [Field L] (v : InfinitePlace K) (w : InfinitePlace L)
+
+@[simp]
+theorem comap_apply (f : K →+* L) (x : K) :
+    w.comap f x = w (f x) := rfl
+
+variable {v w} in
+theorem comp_of_comap_eq {f : K →+* L} (h : w.comap f = v) (x : K) :
+    w (f x) = v x := by
+  simp [← h]
+
+/-- Class encoding the fact that an infinite place is real. -/
+class Real where
+  isReal' : v.IsReal
+
+theorem Real.isReal [v.Real] : v.IsReal := Real.isReal'
+
+/-- Class encoding the fact that an infinite place is complex. -/
+class Complex where
+  isComplex' : v.IsComplex
+
+theorem Complex.isComplex [v.Complex] : v.IsComplex := Complex.isComplex'
+
+theorem not_real_iff_complex : ¬v.Real ↔ v.Complex := by
```

Comment:
If you're going to have lemmas like this then maybe you should prove v.Real <-> v.IsReal and the same for Complex, and then just rewrite those and `not_isReal_iff_isComplex` -- no need to split into two cases. But if you want lemmas like this then why make `NumberField.InfinitePlace.Real` at all, that's not usually how you use classes.